library/Vanilla/Models/SSOModel.php

```diff
@@ -0,0 +1,319 @@
+<?php
+/**
+ * @copyright 2009-2017 Vanilla Forums Inc.
+ * @license http://www.opensource.org/licenses/gpl-2.0.php GNU GPL v2
+ */
+
+namespace Vanilla\Models;
+
+use Gdn_Configuration;
+use Gdn_Session;
+use UserModel;
+use Vanilla\AddonManager;
+use Vanilla\Utility\CapitalCaseScheme;
+
+/**
+ * Class SSOModel
+ */
+class SSOModel {
+
+    /** @var AddonManager */
+    private $addonManager;
+
+    /** @var Gdn_Configuration */
+    private $config;
+
+    /** @var CapitalCaseScheme */
+    private $capitalCaseScheme;
+
+    /** @var  \Gdn_Session */
+    private $session;
+
+    /** @var UserModel */
+    private $userModel;
+
+    /**
+     * SSOModel constructor.
+     *
+     * @param AddonManager $addonManager
+     * @param Gdn_Configuration $config
+     * @param Gdn_Session $session
+     * @param UserModel $userModel
+     */
+    public function __construct(
+        AddonManager $addonManager,
+        Gdn_Configuration $config,
+        Gdn_Session $session,
+        UserModel $userModel
+    ) {
+        $this->addonManager = $addonManager;
+        $this->capitalCaseScheme = new CapitalCaseScheme();
+        $this->config = $config;
+        $this->session = $session;
+        $this->userModel = $userModel;
+    }
+
+    /**
+     * Create a user from the supplied SSOInfo.
+     *
+     * @param SSOInfo $ssoInfo
+     * @return array|false The user of false on failure.
+     */
+    public function createUser(SSOInfo $ssoInfo) {
+        $email = $ssoInfo->coalesce('email');
+        $name = $ssoInfo->coalesce('name');
+
+        if (!$email && !$name) {
+            return false;
+        }
+
+        $userInfo = [
+            'Name' => $name,
+            'Email' => $email,
+            'Password' => randomString('32', 'Aa0!'),
```

Comment:
It'd be better if the length was a class constant.

Reply:
`randomString` just passes things through to `betterRandomString`, so let's go with that.

Reply:
It is used in one place and is not relevant (the hashmethod is random so any arbitrary number will do). No need to expose this as a class constant.